neps/nep-0488.md

```diff
@@ -0,0 +1,1537 @@
+
+The scalar value is encoded as a big-endian `[u8;32]`. All possible bytes combination is allowed. Encoding is similar with BN254 implementation in nearcore[^50], but `big-endian` encoding is used instead of `little-endian` as in EIP-2537[^15].
+
+#### Fields elements Fp
+
+The value from $F_p$ is encoded as a big-endian `[u8; 48]`. Only values less than `p` are allowed. If the value is equal to or bigger than `p` the error should be returned.
+
+The rule of encoding is consistent with zkcrypto[^53], with implementation in milagro lib[^29].
+
+#### Extension fields elements Fp2
+
+The $q \in F_{p^{2}}$ could be written as $q = c_0 + c_1 v$, where $c_0, c_1 \in F_p$ The element from $F_{p^2}$ encoded in `[u8; 96]` as a bytes’ concatenation of $c_1$ and $c_0$. The $c_1$ and $c_0$ are encoded by the rule described in the previous section.
+
+$q \in F_{p^2}$, $q = c_0 + c_1 v$ encoded as `[u8; 96]`:
+
+- $c_1 \in F_p$ `[u8; 48]`
+- $c_0 \in F_p$ `[u8; 48]`
+
+The rule of encoding is consistent with zkcrypto[^53] and with implementation in milagro lib[^29].
+
+#### Uncompressed points on curve E(Fp)
+
+The points on the curve represent by affine coordinates: `(x: Fp, y: Fp)`. The elements from $E(F_p)$ encoded in `[u8; 96]` as bytes’ concatenation of `x` and `y` point coordinates, where $x, y \in F_p$. The $x, y$ are encoded according to the rules described in the section “Fields elements $F_p$” .
+
+$E(F_p)$  is encoded as `[u8; 96]`:
+
+- $x \in F_p$ `[u8; 48]`
+- $y \in F_p$ `[u8; 48]`
+
+*The second bit* is used to mark a point on infinity, if the second bit is set to 1 — it is an infinity point.
+
+Encoding point on infinity:
+
+```bash
+let x: [u8; 96] = [0; 96];
+x[0] = x[0] | 0x40;
+```
+
+The rule of encoding is consistent with zkcrypto[^53] and with implementation in milagro lib[^29].
+
+#### Compressed points on curve E(Fp)
+
+The points on the curve represent by affine coordinates: `(x: Fp, y: Fp)`. The elements from $E(F_p)$ in compressed form are encoded in `[u8; 48]` as *big-endian* encoded $x \in F_p$. The $y$  coordinate can be detected by the formula: $y = \pm \sqrt{x^3 + 4}$.
+
+*The first bit* should be set as 1. This bit indicates that point is encoded in compressed form.
+
+*The second bit* is used to mark a point on infinity, if the second bit is set to 1 — it is an infinity point.
+
+To represent the sign of the $y$ *the third bit* of x encoding is used. If the first bit = 0 the $y$ is positive, if the first bit = 1 the $y$ is negative. We will consider the number positive the smallest between $y$  and $-y = p - y$.
+
+The $x \in F_p$ encoded as `[u8; 48]` bytes according to the rules from section  “Extension fields elements $F_{p}$” .
+
+The point on $E(F_p)$ with negative $y$ coordinate encoded as `[u8; 48]` bytes:
+
+```rust
+let x: [u8; 48] = encodeFp(x)
+x[0] = x[0] | 0x80;
+x[0] = x[0] | 0x20;
+```
+
+For encoding point of infinity:
+
+```rust
+let x: [u8; 48] = [0; 48];
+x[0] = x[0] | 0x80;
+x[0] = x[0] | 0x40;
+```
```

Comment:
Elliptic curve may have a single point at infinity (not positive or negative, just infinity). This code would produce encoding where all bits other than the first two are zero (because of `[0; 48]`), and I think that the code should reject any other encoding.

Reply:
In this [implementation](https://github.com/zkcrypto/pairing/blob/34aa52b0f7bef705917252ea63e5a13fa01af551/src/bls12_381/README.md) which is used in several places, they also state that, if the bit for the point is at infinity is set, then the remaining bits of the group element's encoding should be set to zero, which matches the suggestion above.

Reply:
@Ekleog-NEAR Yes, all other bits should be set to 0.  I will write it more explicitly and also will add a test case with an incorrect infinity encoding.

Reply:
fix: https://github.com/near/NEPs/pull/488/commits/90cb55f4a0dbd07f7ff794939197cddd3e36c5b6